### PR TITLE
feature : support batch update and delete with multiple sql 

### PR DESCRIPTION
**Pull request by wangwei-ying**
#### support batch update and delete with multiple sql , not support insert multiple sql 
```sql
update account_tbl set money = money - ? where user_id = ?;;update account_tbl set money = money - ? where user_id = ?
```

### Ⅱ. Does this pull request fix one issue?

#1784 

### Ⅲ. Why don't you add test cases (unit test/integration test)? 
had add some simple test cases;

### Ⅳ. Describe how to verify it
use seata-simples verify it

### Ⅴ. Special notes for reviews
in `MulitExecutor` , it  handler `UpdateExecutor` and `DeleteExecutor`  images build,
and merge  them.

### CODE REVIEW

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/MulitExecutor.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+     * @return the table records
+     * @throws SQLException the sql exception
+     * @see io.seata.rm.datasource.sql.SQLVisitorFactory#getMulti(String, String) validate sqlType
+     */
+    @Override
+    protected TableRecords beforeImage() throws SQLException {
+        for (SQLRecognizer recognizer : sqlRecognizers) {
+            AbstractDMLBaseExecutor<T, S> executor = null;
+            switch (recognizer.getSQLType()) {
+                case UPDATE:
```
Comment (zjinlei) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/MulitExecutor.java]: 1.Mulit-->Multi
2.The performance drops too badly
We can subdivide the scene:
a.Multiple tables and multiple types
b.same tale and multiple types
c.same table and same type (hope to be able to assemble the beforeImages with one sql)

Comment (wangwei-ying) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/MulitExecutor.java]: Thanks,I will subdivide it

**Comment by wangwei-ying** [comment]
@zjinlei Hi,I have a question , in `buildWhereCondition()` ,this method  use  ` or `   to concat batch params , but it will disable index .  
 ` union ` may be better?

**Comment by wangwei-ying** [comment]
@zjinlei done

**Comment by zjinlei** [comment]
please resolve conflict
and add your git association information on the next commit

**Comment by wangwei-ying** [comment]
> please resolve conflict
> and add your git association information on the next commit

done

**Comment by zjinlei** [comment]
please resolve conflicts

**Comment by wangwei-ying** [comment]
> please resolve conflicts

done

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            selectSQLAppender.add(getColumnNameInSQL(keywordChecker.checkAndReplace(column)));
+        }
+        return buildTableRecords(tmeta, selectSQLAppender.toString(), paramAppenderList);
+    }
+
+    @Override
+    protected TableRecords afterImage(TableRecords beforeImage) throws SQLException {
+        if (sqlRecognizers.size() == 1) {
+            UpdateExecutor executor = new UpdateExecutor<>(statementProxy, statementCallback, sqlRecognizers.get(0));
+            return executor.beforeImage();
```
Comment (zjinlei) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java:94]: afterImage

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        StringBuilder prefix = new StringBuilder("SELECT ");
+        if (!tmeta.containsPK(new ArrayList<>(updateColumnsSet))) {
+            prefix.append(getColumnNameInSQL(tmeta.getEscapePkName(getDbType()))).append(", ");
+        }
+        final StringBuilder suffix = new StringBuilder(" FROM ").append(getFromTableInSQL());
+        suffix.append(" WHERE ").append(whereCondition);
+        suffix.append(" FOR UPDATE");
+        final StringJoiner selectSQLAppender = new StringJoiner(", ", prefix, suffix.toString());
+        for (String column : updateColumnsSet) {
+            selectSQLAppender.add(getColumnNameInSQL(keywordChecker.checkAndReplace(column)));
```
Comment (zjinlei) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java:85]: From the user's sql, no need to deal with keywords

Comment (wangwei-ying) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java]: done

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/ExecuteTemplate.java [hunk 4]
```diff
@@ -69,31 +70,36 @@
...
-        if (sqlRecognizer == null) {
-            sqlRecognizer = SQLVisitorFactory.get(
+        if (sqlRecognizers == null) {
+            sqlRecognizers = SQLVisitorFactory.get(
                     statementProxy.getTargetSQL(),
                     statementProxy.getConnectionProxy().getDbType());
         }
         Executor<T> executor;
-        if (sqlRecognizer == null) {
+        if (sqlRecognizers == null || sqlRecognizers.size() == 0) {
```
Comment (jsbxyyx) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/ExecuteTemplate.java:79]: CollectionUtils.isEmpty()

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiExecutor.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        return null;
+    }
+
+
+    @Override
+    protected void prepareUndoLog(TableRecords beforeImage, TableRecords afterImage) throws SQLException {
+        if (beforeImagesMap == null || afterImagesMap == null) {
+            throw new IllegalStateException("images can not be null");
+        }
+        for (SQLRecognizer recognizer : beforeImagesMap.keySet()) {
```
[the remark this answers is not in the document]
Comment (slievrly) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiExecutor.java:114]: update A set xxx where id=1; delete from A  where id=1;

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiExecutor.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+     * Before image table records.  only support update or deleted
+     *
+     * @return the table records
+     * @throws SQLException the sql exception
+     * @see io.seata.rm.datasource.sql.SQLVisitorFactory#getMulti(String, String) validate sqlType
+     */
+    @Override
+    protected TableRecords beforeImage() throws SQLException {
+        //group by sqlType and table
+        multiSqlGroup = sqlRecognizers.stream().collect(Collectors.groupingBy(t -> t.getSQLType().value() + t.getTableName()));
```
[the remark this answers is not in the document]
Comment (zjinlei) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiExecutor.java:68]: @wangwei-ying I have n’t thought about it, please only support all update batches or all delete batches first.

[the remark this answers is not in the document]
Comment (wangwei-ying) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiExecutor.java:68]: @zjinlei OK,I think so.

[the remark this answers is not in the document]
Comment (wangwei-ying) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiExecutor.java:68]: @zjinlei  but  assemble the images always  rearrange sequences 。Is it allowed to do？

[the remark this answers is not in the document]
Comment (zjinlei) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiExecutor.java:68]: Single sql type does not need to use Map, use List directly, why there is sorting problem?

[the remark this answers is not in the document]
Comment (wangwei-ying) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiExecutor.java:68]: We need to merge images  group by  table name.

**Comment by jsbxyyx** [comment]
update test1 set id = ? where id = ?; update test1 set id = ? where id = ?;
?, ? = 1, 1
?, ? = 2, 2

**Comment by jsbxyyx** [comment]
update test1 set id = ? where id = ?; update test1 set id = ? where id = ?;
?, ? = 1, 1
?, ? = 2, 2

Parameter index out of range (1 > number of parameters, which is 0)

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        Set<String> updateColumnsSet = new HashSet<>();
+        StringBuilder whereCondition = new StringBuilder();
+        for (SQLRecognizer recognizer : sqlRecognizers) {
+            sqlRecognizer = recognizer;
+            SQLUpdateRecognizer sqlUpdateRecognizer = (SQLUpdateRecognizer) recognizer;
+            List<String> updateColumns = sqlUpdateRecognizer.getUpdateColumns();
+            updateColumnsSet.addAll(updateColumns);
+            String whereConditionStr = buildWhereCondition(sqlUpdateRecognizer, paramAppenderList);
+            if (StringUtils.isBlank(whereConditionStr)) {
+                whereCondition = new StringBuilder();
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java:67]: update A set a1=xx, a2=xx where id=1;
update A set a1=xx;
update A set a3=xx where id=2;

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+    @Override
+    protected TableRecords afterImage(TableRecords beforeImage) throws SQLException {
+        if (sqlRecognizers.size() == 1) {
+            UpdateExecutor executor = new UpdateExecutor<>(statementProxy, statementCallback, sqlRecognizers.get(0));
+            return executor.afterImage(beforeImage);
+        }
+        TableMeta tmeta = getTableMeta(beforeImage.getTableName());
+        if (beforeImage == null || beforeImage.size() == 0) {
+            return TableRecords.empty(getTableMeta());
```
Comment (jsbxyyx) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java:106]: update test1 set id = 1 where id = 1; update test1 set name = '11' where name = '11';

Exception: NPE

Comment (wangwei-ying) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java]: done

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/ExecuteTemplate.java (hunk 4 above)
Comment (jsbxyyx) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/ExecuteTemplate.java:79]: CollectionUtils.isEmpty()?

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        return buildTableRecords(tmeta, selectSQLAppender.toString(), paramAppenderList);
+    }
+
+    @Override
+    protected TableRecords afterImage(TableRecords beforeImage) throws SQLException {
+        if (sqlRecognizers.size() == 1) {
+            UpdateExecutor executor = new UpdateExecutor<>(statementProxy, statementCallback, sqlRecognizers.get(0));
+            return executor.afterImage(beforeImage);
+        }
+        if (beforeImage == null || beforeImage.size() == 0) {
```
Comment (jsbxyyx) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java:104]: CollectionUtils.isEmpty()?

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/ExecuteTemplate.java (hunk 4 above)
Comment (wangwei-ying) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/ExecuteTemplate.java]: fine

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java (hunk 9 above)
Comment (wangwei-ying) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java:104]: beforeImage is not a collection

Comment (jsbxyyx) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java:104]: sorry.

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        if (sqlRecognizers.size() == 1) {
+            UpdateExecutor executor = new UpdateExecutor<>(statementProxy, statementCallback, sqlRecognizers.get(0));
+            return executor.beforeImage();
+        }
+        final TableMeta tmeta = getTableMeta(sqlRecognizers.get(0).getTableName());
+
+        final ArrayList<List<Object>> paramAppenderList = new ArrayList<>();
+        Set<String> updateColumnsSet = new HashSet<>();
+        StringBuilder whereCondition = new StringBuilder();
+        boolean selectAll = false;
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/MultiUpdateExecutor.java:60]: variables need to be clear.

**Pull request merged**